### PR TITLE
Implement ngx.shared.DICT.decr method

### PR DESCRIPTION
**Pull request by p0pr0ck5**
Corresponding with lua-nginx-module PR #943.

### CODE REVIEW

**Comment by p0pr0ck5** [comment]
Added tests for the new decr(), but obviously this is failing because the upstream master doesn't have this ffi function in `ngx_http_lua_shdict.c`. I didn't want to update .travis.yml with a patch if it wasn't necessary; once the openresty change is merged these tests will pass.

File: t/shdict.t [hunk 1]
```diff
@@ -999,3 +999,179 @@ number expected, got string
...
 [alert]
 [crit]
+
+
+
+=== TEST 31: decr int
+--- http_config eval: $::HttpConfig
+--- config
+    location = /t {
+        content_by_lua '
```
Comment (agentzh) [inline t/shdict.t]: Newly added test cases should use `*_by_lua_block`. Please fix other test cases as well. Thank you.

Comment (lordnynex) [inline t/shdict.t]: @p0pr0ck5 Added the above changes in https://github.com/p0pr0ck5/lua-resty-core/pull/1

Comment (p0pr0ck5) [inline t/shdict.t]: ^ merged into my branch

File: lib/resty/core/shdict.lua [hunk 2]
```diff
@@ -380,6 +383,16 @@ local function shdict_incr(zone, key, value, init)
 end
 
 
+local function shdict_incr(zone, key, value, init)
+    return shdict_incr_helper(zone, 0x0001, key, value, init)
+end
+
+
+local function shdict_decr(zone, key, value, init)
+    return shdict_incr_helper(zone, 0x0002, key, value, init)
```
Comment (agentzh) [inline lib/resty/core/shdict.lua]: Can we just reuse the existing C API and negate `value` here? If we want to reject decrementing a negative value, we should do the check just here. But frankly I don't think it's worth it for an extra branch.

**Pull request closed without merging**